### ISSUE TITLE
Storage Manager not showing the temperature of the drive

### ISSUE THREAD

**Issue by AuthorShin** [IceWhaleTech/CasaOS#295]
Hello
Storage Manager is not showing the temperature of the drive and instead just shows `Temp: N/A`.

**Comment by Bira-oliverio** [comment]
I have the same issue, the temperature only appears for the second drive, but if I run the S.M.A.R.T. diagnostic via command line the temperature of the first drive is shown.

**Comment by AuthorShin** [comment]
@tigerinus Here you go:
<img src="https://i.ibb.co/VNxhrcB/Screenshot-2022-06-16-081353.png" alt="got the HDD temperature with smartctl -a /dev/sda | grep -i temperature">

**Comment by Bira-oliverio** [comment]
```
$ sudo smartctl -a /dev/sda | grep Temperature
194 Temperature_Celsius     0x0002   230   230   000    Old_age   Always       -       26 (Min/Max 10/51)
```

**Comment by tigerinus** [comment]
> ```
> $ sudo smartctl -a /dev/sda | grep Temperature
> 194 Temperature_Celsius     0x0002   230   230   000    Old_age   Always       -       26 (Min/Max 10/51)
> ```

The challenge with `smartctl` is it's a blocking operation that has a performance impact when it is grabbing every single detail at once before the `grep`. We don't want to refresh the hdd temperature every minute in this way.

Not sure if there is a more programmatic approach.

**Comment by tigerinus** [comment]
Closing as it's not feasible for now, but it is still in our radar.

**Comment by AuthorShin** [comment]
@tigerinus I don't understand what you mean by "blocking operation" also how much "performance impact" it has.
P.S. You can use something like "Node exporter" or other similar things to gather the data.
It's not going to have much of a load if you set the data gathering to every 5 minutes (very suitable for hard drives).
Although some people might say it's not updating very frequent but it's better to be on 5 min than not having the HDD temp at all. also you can add an option to select between 1 min / 5 min or even other options with a disclaimer that on lower intervals it might add more load and impact the device performance.

**Comment by tigerinus** [comment]
`sudo smartctl -a /dev/sda | grep Temperature` is something we try to avoid building into our code. It greps the needed information out of an output that includes other unneeded information. Plus that the line it greps for is not something that can be parsed deterministically. We prefer to obtain temperature information from a reliable API.

Last time I checked, there is not a great Go library for S.M.A.R.T protocol, but I could be wrong. If you know anything that we could use, please share.

> P.S. You can use something like "Node exporter" or other similar things to gather the data.

Do you mind share some sample code snippets on this?

**Comment by AuthorShin** [comment]
@tigerinus Sorry for the late reply.
You can find far more than some code snippets [here](https://github.com/MauriceNino/dashdot).

**Comment by Xarth-Mai** [comment]
The issue still persists. The temperature of my second hard drive is displayed as N/A, but when I run smartctl -a -n standby /dev/sda -j, I get the correct temperature result. I also checked the CasaOS code, and theoretically, it should be working fine, but it’s not. Additionally, it was working correctly for a while after I first installed CasaOS.
![P](https://github.com/user-attachments/assets/25425d68-742a-4360-ae60-8b12df1dc219)
![P](https://github.com/user-attachments/assets/48f26b73-6f2b-4a7b-9287-5062bd850d3b)
